### ISSUE TITLE
Why does Libtorch use more memory than Pytorch does in Python？

### ISSUE THREAD

**Issue by qingfengmingyue** [pytorch/pytorch#40133]
https://github.com/pytorch/pytorch/issues/39927#issue-637501953

**Comment by malfet** [comment]
Can you please give more details (and perhaps reopen the original issue rather than moving it into a new one)
I.e. can you post links to Python (I assume it's https://github.com/xinntao/ESRGAN ) and C++ code that you are comparing memory footprint of?

**Comment by qingfengmingyue** [comment]
C++ code:
int main(int argc, const char* argv[]) {

    static torch::jit::script::Module module = torch::jit::load("RRDB_ESRGAN_x4_000.pt");
    module.to(at::kCUDA);

    std::vector<torch::jit::IValue> inputs;
    inputs.push_back(torch::ones({2, 3, 200, 200}).to(at::kCUDA));

    // Execute the model and turn its output into a tensor.
    auto output = module.forward(inputs).toTensor();

    std::cout << output.slice(/*dim=*/1, /*start=*/0, /*end=*/5) << '\n';
    std::cout << "over" << '\n';
 }

ubuntu18.04 ，Directly predict in two ways, on the 2080TI libtorch can only predict 200X200, and pytorch can predict 800X800 size

**Comment by flowvision** [comment]
I also had an issue where the the model in C++ took way more GPU memory than in Python.

For me the problem was that in Python the code for the inference was surrounded by a
`with torch.no_grad():` which I overlooked in the C++ version.

Try to use `torch::NoGradGuard` in your code as follows:

```
int main(int argc, const char* argv[]) {

    static torch::jit::script::Module module = torch::jit::load("RRDB_ESRGAN_x4_000.pt");
    module.to(at::kCUDA);

    torch::NoGradGuard noGrad;

    std::vector<torch::jit::IValue> inputs;
    inputs.push_back(torch::ones({2, 3, 200, 200}).to(at::kCUDA));

    // Execute the model and turn its output into a tensor.
    auto output = module.forward(inputs).toTensor();

    std::cout << output.slice(/*dim=*/1, /*start=*/0, /*end=*/5) << '\n';
    std::cout << "over" << '\n';
}
```

This solved the problem for me. Let me know I this also works for you.
